src/lxc/lxccontainer.c

```diff
@@ -4698,6 +4719,299 @@ static bool do_lxcapi_restore(struct lxc_container *c, char *directory, bool ver
 
 WRAP_API_2(bool, lxcapi_restore, char *, bool)
 
+static int ensure_dummy_dir(const char *dest)
+{
+	int ret;
+	struct stat sb;
+
+	ret = stat(dest, &sb);
+	if (ret == 0) {
+		if (S_ISDIR(sb.st_mode))
+			return 0;
+
+		ret = unlink(dest);
+		if (ret < 0) {
+			SYSERROR("Failed to remove \"%s\"", dest);
+			return -1;
+		}
+	}
+
+	ret = mkdir(dest, 0000);
+	if (ret < 0) {
+		SYSERROR("Failed to create directory \"%s\"", dest);
+		return -1;
+	}
+
+	return 0;
+}
```

Comment:
These two helpers are quite large and also quite similar. I wonder if maybe one suffices:
```C
static int ensure_dummy(const char *dest, mode_t mode)
{
	int ret;
	struct stat sb;

	ret = stat(dest, &sb);
	if (ret == 0) {
		if (!S_ISDIR(sb.st_mode) == !S_ISDIR(mode))
			return 0;

		ret = S_ISDIR(sb.st_mode) ? rmdir(dest) : unlink(dest);
		if (ret < 0) {
			SYSERROR("Failed to remove \"%s\"", dest);
			return -1;
		}
	}

	ret = S_ISDIR(mode) ? mkdir(dest, 0000) : mknod(dest, S_IFREG, 0);
	if (ret < 0) {
		SYSERROR("Failed to create destination \"%s\"", dest);
		return -1;
	}

	return 0;
}
```

At the same time I wonder if we should even bother with the verbose checking and just shorten it to:
```C
static int ensure_dummy(const char *dest, mode_t mode)
{
	int ret;
	struct stat sb;

	ret = stat(dest, &sb);
	if (ret == 0) {
		if (!S_ISDIR(sb.st_mode) == !S_ISDIR(mode))
			return 0;

		// If neither works, mount mkdir/mknod will fail anyway
		(void)rmdir(dest);
		(void)unlink(dest);
	}

	ret = S_ISDIR(mode) ? mkdir(dest, 0000) : mknod(dest, S_IFREG, 0);
	if (ret < 0) {
		SYSERROR("Failed to create destination \"%s\"", dest);
		return -1;
	}

	return 0;
}
```

or even just

```C
static int ensure_dummy(const char *dest, mode_t mode)
{
	int ret;
	struct stat sb;

	(void)rmdir(dest);
	(void)unlink(dest);
	if (S_ISDIR(mode))
		ret = mkdir(dest, 0000);
	else
		ret = mknod(dest, S_IFREG, 0);
	if (ret < 0) {
		SYSERROR("Failed to create destination \"%s\"", dest);
		return -1;
	}

	return 0;
}
```

Simply because AFAICT the shared mount directory is supposed to be internal to lxc and whether the removal of a previously different type of mount point entry failed, or simply its creation, won't usually point out any particular error that wouldn't be obvious from a failing mkdir/mknod already anyway?

Reply:
I prefer your last version, though in terms of coding style I'd prefer if this were:
```C
static int ensure_dummy(const char *dest, mode_t mode)
{
	int ret;

	(void)remove(dest);

	if (S_ISDIR(mode))
		ret = mkdir(dest, 0000);
	else
		ret = mknod(dest, S_IFREG | 0000, 0);
	if (ret < 0) {
		SYSERROR("Failed to create dummy mount target \"%s\"", dest);
		return -1;
	}

	return 0;
}
```

Reply:
Ah yes, there's `remove()`, as well :-)

Reply:
Actually, in this case `ensure_dummy` is rather small, and it's probably easier to make it just a part of the `create_mount_target`.